ftw/upgrade/command/__init__.py

```diff
@@ -142,7 +144,25 @@ def __call__(self):
         args = self.parser.parse_args()
         configure_logging(args)
         setattr(args, 'parser', self.parser)
-        args.func(args)
+        if getattr(args, 'all_sites', False):
+            info = {}
+            while True:
+                try:
+                    with capture() as out:
+                        args.func(args)
+                    output = out.getvalue().strip()
+                    if getattr(args, 'json', False):
+                        # "[]" must be turned into [].
+                        output = json.loads(output)
+                    info[args.picked_site] = output
+                except StopIteration:
+                    break
+            if getattr(args, 'json', False):
+                print(json.dumps(info))
+            else:
+                print(info)
```

Comment:
The non-JSON output is now a python dict:
With all sites:
```sh
% ./bin/upgrade list --profiles -A
INFO: Acting on site /Plone
INFO: Acting on site /blubber
{u'/blubber': u'Installed profiles:\nID:                                 Title:                          Versions (DB/FS):  \nProducts.CMFEditions:CMFEditions    CMFEditions                     4 / 4              \nProducts.TinyMCE:TinyMCE            TinyMCE Editor Support          7 / 7              \nplone.app.querystring:default       Querystring parser/builder      5 / 5              \nplone.app.jquerytools:default       Plone JQuery Tools Integration  3 / 3              \nplone.app.discussion:default        Plone Discussions               101 / 101          \nplone.app.jquery:default            Plone JQuery Integration        3 / 3              \nplonetheme.sunburst:default         Sunburst Theme                  3 / 3', u'/Plone': u'Installed profiles:\nID:                                 Title:                          Versions (DB/FS):  \nProducts.CMFEditions:CMFEditions    CMFEditions                     4 / 4              \nProducts.TinyMCE:TinyMCE            TinyMCE Editor Support          7 / 7              \nplone.app.querystring:default       Querystring parser/builder      5 / 5              \nplone.app.jquerytools:default       Plone JQuery Tools Integration  3 / 3              \nplone.app.discussion:default        Plone Discussions               101 / 101          \nplone.app.jquery:default            Plone JQuery Integration        3 / 3              \nplonetheme.sunburst:default         Sunburst Theme                  3 / 3'}
```
With one site for comparison:
```sh
% ./bin/upgrade list --profiles -L
Installed profiles:
ID:                                 Title:                          Versions (DB/FS):
Products.CMFEditions:CMFEditions    CMFEditions                     4 / 4
Products.TinyMCE:TinyMCE            TinyMCE Editor Support          7 / 7
plone.app.querystring:default       Querystring parser/builder      5 / 5
plone.app.jquerytools:default       Plone JQuery Tools Integration  3 / 3
plone.app.discussion:default        Plone Discussions               101 / 101
plone.app.jquery:default            Plone JQuery Integration        3 / 3
plonetheme.sunburst:default         Sunburst Theme                  3 / 3
```


---


We could either change this print statement so that we print the `info` in a better format, or we could remove it here and print the info directly above if we have no json argument, for example by extending the `if` above:
```diff
diff --git a/ftw/upgrade/command/__init__.py b/ftw/upgrade/command/__init__.py
index 491f0e4..8818b8c 100644
--- a/ftw/upgrade/command/__init__.py
+++ b/ftw/upgrade/command/__init__.py
@@ -154,13 +154,13 @@ class UpgradeCommand(object):
                     if getattr(args, 'json', False):
                         # "[]" must be turned into [].
                         output = json.loads(output)
-                    info[args.picked_site] = output
+                        info[args.picked_site] = output
+                    else:
+                        print output
                 except StopIteration:
                     break
             if getattr(args, 'json', False):
                 print(json.dumps(info))
-            else:
-                print(info)
         else:
             args.func(args)
```

---

What do you think?